src/utils.cc

```diff
@@ -140,7 +140,10 @@ std::string realPath(const std::string &path) {
 bool normalizeFolder(std::string &path) {
   for (auto &[root, real] : g_config->workspaceFolders)
     if (real.size() && llvm::StringRef(path).startswith(real)) {
-      path = root + path.substr(real.size());
+      // extra memory allocation may happen when string concatenates
+      path.clear();
+      path.append(root);
+      path.append(path.substr(real.size()));
```

Comment:
```suggestion
      path.append(path, real.size());
```
no need for the substr temporary, append has an overload to append a sub-string directly